# ML-08 — Capstone Modeling Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/inshrah-malik/inshrah-flyrank-ml-internship/blob/main/work/notebooks/w05_model.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

Load  dataset

In [5]:
from datasets import load_dataset

ds = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_daily_performance",
    split="train[:10000]"
)

df = ds.to_pandas()

README.md:   0%|          | 0.00/3.04k [00:00<?, ?B/s]

Resolving data files:   0%|          | 0/18 [00:00<?, ?it/s]

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 4.41MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 2.62MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 90.3MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 19.6kB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 72.0MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  624kB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 1.45MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 21.6MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 86.2MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 8.93MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.29MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  124MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 7.12MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 89.0MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.22MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  134MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  149MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  146MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/78835655 [00:00<?, ? examples/s]

CTR

In [6]:
df["ctr"] = df["gsc_clicks"] / df["gsc_impressions"]

In [7]:
df["ctr"] = (
    df["gsc_clicks"] /
    df["gsc_impressions"].replace(0,1)
)

Selected Features

In [8]:
features = [

    "gsc_impressions",

    "gsc_avg_position",

    "ga4_pageviews",

    "ga4_sessions",

    "ga4_users",

    "sessions_organic",

    "sessions_direct",

    "sessions_social",

    "scroll_events"

]

In [21]:
print(features)

['gsc_impressions', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'sessions_organic', 'sessions_direct', 'sessions_social', 'scroll_events']


In [9]:
target = "ctr"

In [10]:
model_df = df[features+[target]].dropna()

Train/Test Split

In [11]:
from sklearn.model_selection import train_test_split

X = model_df[features]

y = model_df[target]

X_train,X_test,y_train,y_test = train_test_split(

    X,

    y,

    test_size=0.2,

    random_state=42

)

Baseline Model

In [12]:
import numpy as np
from sklearn.metrics import mean_absolute_error

baseline_prediction = np.repeat(
    y_train.mean(),
    len(y_test)
)

baseline_mae = mean_absolute_error(
    y_test,
    baseline_prediction
)

baseline_mae

0.019303336835830548

Train Linear Regression

In [13]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train,y_train)

pred_lr = lr.predict(X_test)

In [14]:
from sklearn.metrics import mean_absolute_error

mae_lr = mean_absolute_error(
    y_test,
    pred_lr
)

mae_lr

0.019508072474219908

Decision Tree

In [15]:
from sklearn.tree import DecisionTreeRegressor

tree = DecisionTreeRegressor(

    random_state=42,

    max_depth=5

)

tree.fit(X_train,y_train)

pred_tree = tree.predict(X_test)

mae_tree = mean_absolute_error(
    y_test,
    pred_tree
)

mae_tree

0.01809402016668154

Random Forest

In [16]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(

    random_state=42,

    n_estimators=100

)

rf.fit(X_train,y_train)

pred_rf = rf.predict(X_test)

mae_rf = mean_absolute_error(
    y_test,
    pred_rf
)

mae_rf

0.019130905108834476

Compare Models

In [17]:
import pandas as pd

results = pd.DataFrame({

    "Model":[

        "Baseline",

        "Linear Regression",

        "Decision Tree",

        "Random Forest"

    ],

    "MAE":[

        baseline_mae,

        mae_lr,

        mae_tree,

        mae_rf

    ]

})

results

,Model,MAE
0,Baseline,0.019303
1,Linear Regression,0.019508
2,Decision Tree,0.018094
3,Random Forest,0.019131


In [18]:
importance = pd.DataFrame({

    "Feature":features,

    "Importance":rf.feature_importances_

})

importance.sort_values(
    "Importance",
    ascending=False
)

,Feature,Importance
1,gsc_avg_position,0.699065
0,gsc_impressions,0.300935
2,ga4_pageviews,0.000000
3,ga4_sessions,0.000000
4,ga4_users,0.000000
5,sessions_organic,0.000000
6,sessions_direct,0.000000
7,sessions_social,0.000000
8,scroll_events,0.000000


Error Analysis

In [19]:
comparison = pd.DataFrame({

    "Actual":y_test,

    "Predicted":pred_rf

})

comparison.head(20)

,Actual,Predicted
5344,0.000000,0.003128
7444,0.000000,0.000000
1731,0.000000,0.021709
8719,0.000000,0.000000
4521,0.000000,0.019846
7453,0.000000,0.000000
576,0.000000,0.000000
7428,0.000000,0.011250
5577,0.000000,0.002800
439,0.000000,0.019831


In [20]:
comparison["Error"] = abs(comparison["Actual"] - comparison["Predicted"])
comparison.sort_values("Error", ascending=False).head(10)

,Actual,Predicted,Error
4687,1.000000,0.000000,1.000000
3463,1.000000,0.100374,0.899626
6658,1.000000,0.203093,0.796907
1002,0.500000,0.000000,0.500000
5407,0.500000,0.000000,0.500000
1418,0.500000,0.051764,0.448236
7474,0.375000,0.011956,0.363044
393,0.333333,0.000000,0.333333
427,0.333333,0.000000,0.333333
8507,0.333333,0.000000,0.333333


## 1. Method choice and why

*Which method from the toolkit, and why it fits your lane.*

I selected Random Forest Regression because the task is to predict CTR, which is a continuous value. Random Forest can model non-linear relationships and interactions between search performance features while being more robust than a single decision tree.

## 2. Split design

*Grouped by client? Time-aware? Say why this split is honest for your question.*

I used an 80/20 train-test split with random_state=42 so every model was evaluated on the same test set. The baseline and all models used identical training and testing data for a fair comparison.

## 3. Train + compare vs my baseline

*Same data, same metric, same split as your Week-4 baseline. Show the table.*

I trained three regression models (Linear Regression, Decision Tree, and Random Forest) and compared them with the Week 4 baseline using the same train-test split and the Mean Absolute Error (MAE) metric. The Decision Tree model achieved the lowest MAE of 0.018094, outperforming the baseline MAE of 0.019303. Linear Regression had the highest MAE (0.019508), while Random Forest achieved an MAE of 0.019131. Based on these results, the Decision Tree model provided the most accurate CTR predictions on this dataset.

## 4. Errors and interpretation

*Where is the model wrong? What does it lean on? A short error analysis beats a big metric table.*

The Decision Tree model produced the smallest prediction errors overall, indicating that it captured the relationship between the selected search performance features and CTR better than the other models. Although the improvement over the baseline is relatively small, it demonstrates that the model learned useful patterns from the data. Some prediction errors are still expected because CTR can be influenced by factors that were not included in the selected features.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.